### ISSUE TITLE
Hotplug Storage Support

### ISSUE THREAD

**Issue by Continous** [Nexus-Mods/NexusMods.App#1148]
# User story

**As a** User

**I want to** install and manage games on removable storage devices.

**So that** I can save internal drive space.

## Requirements
* The ability to gracefully remove a storage device with a managed game on it.
* The ability to gracefully insert a storage device with a managed game on it.
* Some method of communicating to the user that a game is on a removable device, and that that device is/is not inserted.
* Maybe extend this to network devices?

## Design
_(Any designs required for this story)_

## DOD
A user should be able to install, manage, and play a game on storage device A. They should then be able to remove storage device A, insert storage device B, install, manage, and play a game on storage device B. After this they should be able to freely remove or insert storage device A or B to resume installing, managing, or playing any games prior installed, managed, or played on those devices. Ideally this would also include any changes from another computer. This functionality may also be useful for network storage.

## Scenario 1
**Given** I manage an installation of Fallout 4 for my Steamdeck.

**When** this installation is on an SD card.

**Then** I should be able to freely insert or remove this SD card.

## Scenario 2

**Given** I manage an installation of Skyrim.

**When** this installation is on a network drive.

**Then** network reliability/reachability issues should be handled gracefully.

**Comment by erri120** [comment]
Did you try this out with the latest version? As far as the App is concerned, it doesn't matter what type of device the game is installed at. As long as the game was installed properly via one of the supported game stores, the App can find the installation and manage it.

**Comment by Continous** [comment]
> Did you try this out with the latest version? As far as the App is concerned, it doesn't matter what type of device the game is installed at. As long as the game was installed properly via one of the supported game stores, the App can find the installation and manage it.

I think you misunderstand. The concern is more about handling the removal of a drive with a managed game on it.

**Comment by erri120** [comment]
> > Did you try this out with the latest version? As far as the App is concerned, it doesn't matter what type of device the game is installed at. As long as the game was installed properly via one of the supported game stores, the App can find the installation and manage it.
> 
> I think you misunderstand. The concern is more about handling the removal of a drive with a managed game on it.

If you manage a game with the App and then move or uninstall the game, I think you'll just get a ton of exceptions in the logs and the UI might crash.

**Comment by Continous** [comment]
> > > Did you try this out with the latest version? As far as the App is concerned, it doesn't matter what type of device the game is installed at. As long as the game was installed properly via one of the supported game stores, the App can find the installation and manage it.
> > 
> > 
> > I think you misunderstand. The concern is more about handling the removal of a drive with a managed game on it.
> 
> If you manage a game with the App and then move or uninstall the game, I think you'll just get a ton of exceptions in the logs and the UI might crash.

Correct. As a user I'd like it not to do that.

**Comment by Pickysaurus** [comment]
This isn't a very common use case. At best the App should probably acknowledge the game is now missing and continue as if it had been removed properly.

**Comment by Continous** [comment]
> This isn't a very common use case. At best the App should probably acknowledge the game is now missing and continue as if it had been removed properly.

I agree that the use case isn't super common at the moment, but with devices like the ROG Ally, Valve Steamdeck, and similar becoming more common I don't think it will stay uncommon.

With that said. I think your proposed solution is a good enough one.

**Comment by Sewer56** [comment]
I think the main stated concern here is not showing games which are currently inaccessible.

---

As far as running newly inserted games, the story is a bit different. It's up to having the external launcher/store configured correctly at that point. Else they'll likely be unable to start the game (due to DRM), even unmodded, in the first place.

**Comment by Al12rs** [comment]
> I think the main stated concern here is not showing games which are currently inaccessible.

Rather than that I would say, that the app should show the game in some capacity and tell the user that it is no longer available and why, as well as allowing the user to potentially redirect the app to the new location where the game was moved (if it was moved for example).

Other mod managers had or have issues with this where if the game is missing they get in a startup lock loop where recovering requires editing or clearing the manager data. We should avoid that kind of issues, allowing users to recover by either fixing the problem or clearing out the orphaned data in the app

**Comment by Sewer56** [comment]
Very true, 100% agreed.
Been a fairly common problem historically.

**Comment by Continous** [comment]
> > I think the main stated concern here is not showing games which are currently inaccessible.
> 
> Rather than that I would say, that the app should show the game in some capacity and tell the user that it is no longer available and why, as well as allowing the user to potentially redirect the app to the new location where the game was moved (if it was moved for example).
> 
> Other mod managers had or have issues with this where if the game is missing they get in a startup lock loop where recovering requires editing or clearing the manager data. We should avoid that kind of issues, allowing users to recover by either fixing the problem or clearing out the orphaned data in the app

It's worth noting that this will likely help to if and whenever someone moves their game installs or has them moved.
